spec.bs

```diff
@@ -512,42 +512,20 @@ To <dfn>validate and convert auction ad config</dfn> given an {{AuctionAdConfig}
-    1. [=Upon fulfillment=] of |auctionConfig|'s [=auction config/per buyer signals=] with |result|:
-      1. Let |pendingException| be null.
-      1. If |result| is not a [=record=] mapping from {{USVString}} to {{any}}, set |pendingException| to
-        a {{TypeError}} and jump to the step labeled
-        <i><a href=#validate-per-buyer-signals-finish>finish</a></i>.
-      1. Set |auctionConfig|'s [=auction config/per buyer signals=] to a new [=ordered map=] whose
-        [=map/keys=] are [=origins=] and whose [=map/values=] are [=strings=].
-      1. [=map/For each=] |key| → |value| of |result|:
-        1. Let |buyer| be the result of [=parsing an origin=] with |key|. If |buyer| is failure,
-          set |pendingException| to a {{TypeError}} and jump to the step labeled
-          <i><a href=#validate-per-buyer-signals-finish>finish</a></i>.
-        1. Let |signalsString| be the result of [=serializing a JavaScript value to a JSON string=], given
-          |value|. If this [=exception/throws=] an exception |e|, set |pendingException| to |e| and jump
-          to the step labeled <i><a href=#validate-per-buyer-signals-finish>finish</a></i>.
-        1. [=map/Set=] |auctionConfig|'s [=auction config/per buyer signals=][|buyer|] to
-          |signalsString|.
-      1. <i id=validate-per-buyer-signals-finish>Finish</i>:
-        1. If |pendingException| is not null:
-          1. Set |auctionConfig|'s [=auction config/per buyer signals=] to failure.
-          1. Decrement |auctionConfig|'s [=auction config/pending promise count=].
-          1. [=exception/Throw=] |pendingException|.
-        1. Otherwise, decrement |auctionConfig|'s [=auction config/pending promise count=].
-    1. [=Upon rejection=] of |auctionConfig|'s [=auction config/per buyer signals=]:
-      1. Set |auctionConfig|'s [=auction config/per buyer signals=] to failure.
+  1. Set |auctionConfig|'s [=auction config/per buyer signals=] to
+    |config|["{{AuctionAdConfig/perBuyerSignals}}"].
+  1. Increment |auctionConfig|'s [=auction config/pending promise count=].
+  1. Let |resolvedAndTypeChecked| be the promise representing performing the following steps [=upon fulfillment=] of |auctionConfig|'s [=auction config/per buyer signals=] with |result|:
+    1. Set |auctionConfig|'s [=auction config/per buyer signals=] to a new [=ordered map=] whose [=map/keys=] are [=origins=] and whose [=map/values=] are [=strings=].
+    1. [=map/For each=] |key| → |value| of |result|:
+      1. Let |buyer| be the result of [=parsing an origin=] with |key|. If |buyer| is failure, throw a {{TypeError}}.
+      1. Let |signalsString| be the result of [=serializing a JavaScript value to a JSON string=], given |value|.
+      1. [=map/Set=] |auctionConfig|'s [=auction config/per buyer signals=][|buyer|] to |signalsString|.
+    1. If the previous steps did not [=exception/throw=] an exception:
       1. Decrement |auctionConfig|'s [=auction config/pending promise count=].

```

Comment:
why do we only decrement `pending promise count` when the previous steps didn't throw? No matter whether the previous steps throw or not, the promise is already resolved, am I understanding it wrong?

Reply:
If they threw, you end up in the reject handler. Basically, if I understood correctly, you end up with:
let resolvedAndTypeChecked = userPromise.then((incoming) =>{
      result = typecheck(incoming);
     // The "upon fullfillment" steps
  })
resolvedAndTypeChecked.catch(() => {
  // The "upon rejection" steps
})

Reply:
... Though I think I dropped one level of .then in there.

Reply:
It's probably more like:
let resolvedAndTypeChecked = userPromise.then((incoming) => {
  return typecheck(incoming);
}).then(result) {
  // The "upon fullfillment" ste[s"
}
resolvedAndTypeChecked.catch(() => {
  // The "upon rejection" steps
})

If I am not confused, any exception thrown by typecheck or "upon fullfillment" will end up in the "upon reject" handler.

Reply:
Your two chunks of code have the same behavior, but the spec is closer to the first.